File: solutions/copy-list-with-random-pointer/copy-list-with-random-pointer.py
// @Title: Copy List with Random Pointer
// @Author: colinjxc
// @Date: 2022-02-05T21:40:08+08:00
// @URL: https://leetcode-cn.com/problems/copy-list-with-random-pointer


"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        curr = head
        node_map = dict()
        while curr:
            node_map[curr] = Node(curr.val)
            curr = curr.next
        curr = head
        
        while curr:
            node = node_map[curr]
            if curr.next:
                node.next = node_map[curr.next]
            if curr.random:
                node.random = node_map[curr.random]
            curr = curr.next
        if head == None:
            return None
        return node_map[head]

File: solutions/copy-list-with-random-pointer/copy-list-with-random-pointer.go
// @Title: Copy List with Random Pointer
// @Author: colinjxc
// @Date: 2022-02-05T21:51:23+08:00
// @URL: https://leetcode-cn.com/problems/copy-list-with-random-pointer


/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Next *Node
 *     Random *Node
 * }
 */

func copyRandomList(head *Node) *Node {
    node_map := make(map[*Node]*Node)
    for curr := head; curr != nil; curr = curr.Next {
        node_map[curr] = &Node{Val:curr.Val}
    }

    for curr := head; curr != nil; curr = curr.Next {
        node := node_map[curr]
        if curr.Next != nil {
            node.Next = node_map[curr.Next]
        }
        if curr.Random != nil {
            node.Random = node_map[curr.Random]
        }
    }
    return node_map[head]
}
